src/lib.rs

```diff
@@ -158,62 +157,82 @@ pub fn create_json(host: &AnalysisHost, crate_name: &str) -> Result<String> {
 
     let root_def = host.get_def(root_id)?;
 
-    fn recur(id: &analysis::Id, host: &AnalysisHost) -> Vec<analysis::Def> {
-        let mut ids = Vec::new();
-        let mut defs = host.for_each_child_def(*id, |id, def| {
-            ids.push(id);
-            def.clone()
-        }).unwrap();
-
-        let child_defs: Vec<analysis::Def> = ids.into_par_iter()
-            .map(|id: analysis::Id| recur(&id, host))
-            .reduce(Vec::default, |mut a: Vec<analysis::Def>,
-             b: Vec<analysis::Def>| {
-                a.extend(b);
-                a
-            });
-        defs.extend(child_defs);
-        defs
-    }
+    fn recur(id: &analysis::Id, host: &AnalysisHost) -> Result<Vec<Document>> {
+        let def = host.get_def(*id)?;
 
-    let mut included: Vec<Document> = Vec::new();
-    let mut relationships: HashMap<String, Vec<Data>> = HashMap::with_capacity(METADATA_SIZE);
+        let mut ids = Vec::new();
+        host.for_each_child_def(*id, |id, _| ids.push(id))?;
 
-    for def in recur(&root_id, host) {
-        let (ty, relations_key) = match def.kind {
-            DefKind::Mod => (String::from("module"), String::from("modules")),
-            DefKind::Struct => (String::from("struct"), String::from("structs")),
-            _ => continue,
+        let ty = match def.kind {
+            DefKind::Mod => String::from("module"),
+            DefKind::Struct => String::from("struct"),
+            _ => String::from("oh no"),
         };
 
         // Using the item's metadata we create a new `Document` type to be put in the eventual
         // serialized JSON.
-        included.push(
-            Document::new()
-                .ty(ty.clone())
-                .id(def.qualname.clone())
-                .attributes(String::from("name"), def.name)
-                .attributes(String::from("docs"), def.docs),
-        );
+        let mut doc = Document::new()
+            .ty(ty.clone())
+            .id(def.qualname.clone())
+            .attributes(String::from("name"), def.name)
+            .attributes(String::from("docs"), def.docs);
 
-        let item_relationships = relationships.entry(relations_key).or_insert_with(
-            Default::default,
-        );
-        item_relationships.push(Data::new().ty(ty).id(def.qualname));
+        let mut docs = Vec::new();
+
+        let child_docs: Vec<Document> = ids.into_par_iter()
+            .map(|id: analysis::Id| recur(&id, host))
+            .reduce(|| Ok(Vec::new()), |a: Result<Vec<Document>>,
+             b: Result<Vec<Document>>| {
+                let mut a = a?;
+                let b = b?;
```

Comment:
you could rewrite this by declaring `|mut a: Result<` for the closure and  then rewrite the body as:

```
a?.extend(b?);
a
```

I think